### PR TITLE
[FIX] website_event: ensure uniqueness for submenu URL

### PR DESCRIPTION
**Pull request by nboulif**
Steps to reproduce:
-------------------

- Install `website_event` module
- Create an event with name `Test Event` and enable `Website Submenu` option
- Click on `Go to Website` stat button
- Edit the `Introduction` page by adding a new block under the event name for example
- Save it and go back to backend
- Create a new event with same name `Test Event` and enable `Website Submenu` option
- Click on `Go to Website` stat button

Issue:
------

The second event will have the edited
`Introduction` page from the first event.

Cause:
------

When creating the menu, since we do not create a
page, the URL retrieved in the process will be
always the same and therefore redirect to the
edited page of the first event (since the last
part of the URL is used to retrieve the right
view).

Solution:
---------

Ensure the URL is unique by using the view key
(who is unique) as last part of the URL.

opw-3945840

### CODE REVIEW

File: addons/website_event/models/event_event.py [hunk 1]
```diff
@@ -376,8 +376,11 @@ def _create_menu(self, sequence, name, url, xml_id, menu_type):
             page_result = self.env['website'].sudo().new_page(
                 name=name + ' ' + self.name, template=xml_id,
                 add_menu=False, ispage=False)
-            url = "/event/" + slug(self) + "/page" + page_result['url']  # url contains starting "/"
-            view_id = page_result['view_id']
+            view_id = page_result["view_id"]
+            view = self.env["ir.ui.view"].browse(view_id)
+            url = (
+                "/event/" + slug(self) + "/page/" + view.key.split(".")[-1]
+            )  # url contains starting "/"
```
Comment (reth-odoo) [inline addons/website_event/models/event_event.py:383]: Seems fine to inline, not sure I get the comment
```suggestion
            url = "event/" + slug(self) + "/page/" + view.key.split(".")[-1]
```

File: addons/website_event/tests/test_event_menus.py [hunk 2]
```diff
@@ -67,3 +67,32 @@ def test_menu_management_frontend(self):
...
+                }
+                for _ in range(2)
+            ]
+        )
+
+        for event_1_menu, event_2_menu in zip(
+            event_1.menu_id.child_id, event_2.menu_id.child_id
+        ):
+            self.assertNotEqual(
+                event_1_menu.url.split("/")[-1], event_2_menu.url.split("/")[-1]
```
Comment (reth-odoo) [inline addons/website_event/tests/test_event_menus.py:97]: Would maybe go one step further and check that the end of the url matches a view each time

File: addons/website_event/models/event_event.py [hunk 3]
```diff
@@ -376,8 +376,11 @@ def _create_menu(self, sequence, name, url, xml_id, menu_type):
             page_result = self.env['website'].sudo().new_page(
```
Comment (reth-odoo) [inline addons/website_event/models/event_event.py:376]: Would say something like
```
the event_page route used to resolve event pages expects URLs to match view names which is not necessarily the case for the returned url of `new_page`
```
in the commit

**Review by reth-odoo — COMMENTED** [review]
Thanks!
Seems good to me, a couple tiny comments otherwise good to go

Already pinging @jeh-odoo for r+

**Comment by jeh-odoo** [comment]
Hello there :wave: 
Funny behavior indeed, normally `new_page()` should call `get_unique_path()`. But since we don't create a website.menu or website.page, the url is never updated to make it unique.

Hum even with this solution we still have some weird behavior as we can access the two pages in each event. I suppose we could have a 404 for one the page instead.
So perhaps we should take into account the `event.event` record to get the `website.event.menu` and to get the view that way in the route for the page ?
Just an idea, as I'm not sure if it's worth the trouble.
What's your take on this @reth-odoo ?

Cheers!

**Comment by reth-odoo** [comment]
@jeh-odoo Ah I see what you mean, indeed would probably be good to check

Seems find to at least double-check that, even though it shouldn't happen under normal circumstances
something like that?
```
page_view = self.env['ir.view'].search([('key', '=', page)])
views = self.env['website.event.menu'].sudo().search([('event_id', '=', event.id)]).view_id
if page_view not in views:
    return request.not_found()
```

Anyway would probably be good to revisit on master at some point to maybe override the page rendering behavior instead of having custom routes doing questionable stuff. We're the only ones to use the `new_page(ispage=False)` feature :smile:

**Comment by nboulif** [comment]
Hello everyone, thanks for the checks :pray: 

Regarding the fix in the route, do you meant it in stable or in master?
Because if we applied it now, people who have two events already created will be redirected to not found page on the second event? :thinking: 
Not saying that is worse than having a redirection to a page of another event :sweat_smile:  but I'm not sure what's best in this case?

What do you think?

**Comment by jeh-odoo** [comment]
Indeed, I prefer to have still something to display for this case so we could do something like this:
```python
        view = request.env['website.event.menu'].sudo().search([
            ('event_id', '=', event.id),
            ('view_id.key', 'ilike', page)]).view_id
        if not view:
            return request.not_found()

        try:
            # Every event page view should have its own SEO.
            values['seo_object'] = view
```
Seems like we need an `ilike` for previous views as the view key is correctly made unique.
Could you give it a try ?

Cheers!

File: addons/website_event/controllers/main.py [hunk 4]
```diff
@@ -149,9 +149,19 @@ def event_page(self, event, page, **post):
         if '.' not in page:
             page = 'website_event.%s' % page
 
+        view = (
+            request.env["website.event.menu"]
+            .sudo()
+            .search([("event_id", "=", event.id), ("view_id.key", "ilike", page)])
+            .view_id
+        )
```
Comment (jeh-odoo) [inline addons/website_event/controllers/main.py:157]: Perhaps a bit too much on the multilines here 🙂 Something like this perhaps:
```suggestion
        view = request.env["website.event.menu"].sudo().search([
            ("event_id", "=", event.id), ("view_id.key", "ilike", page)]).view_id
```

**Review by jeh-odoo — COMMENTED** [review]
Hey @nboulif 
Just two last picky comments to reduce a bit the diff 🥺 
And then we should be good to go 🚀 
Cheers!

File: addons/website_event/controllers/main.py (hunk 4 above)
Comment (nboulif) [reply addons/website_event/controllers/main.py]: My bad, it's `Ruff` that automatically format it on save. Updated :+1:

File: addons/website_event/controllers/main.py [hunk 5]
```diff
@@ -149,9 +149,19 @@ def event_page(self, event, page, **post):
...
+        if not view:
+            return request.not_found()
+
         try:
             # Every event page view should have its own SEO.
-            values['seo_object'] = request.website.get_template(page)
+            page = view.key
+            values["seo_object"] = request.website.get_template(page)
```
[the remark this answers is not in the document]
Comment (nboulif) [reply addons/website_event/controllers/main.py]: Same here (since changed in previous commit). Updated :+1:

**Review by jeh-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**